src/v/kafka/server/usage_manager.cc

```diff
@@ -0,0 +1,386 @@
+      _buckets.at(_current_window).u,
+      [](const usage& acc, const usage& x) { return acc + x; });
+    _current_window = (_current_window + 1) % _buckets.size();
+    _buckets.at(_current_window).reset(now);
+}
+
+std::chrono::seconds usage_manager::accounting_fiber::reset_state(
+  std::vector<usage_window> buckets) {
+    /// called after restart to determine which bucket is the 'current' bucket
+    auto last_window_delta = std::chrono::seconds(0);
+    _current_window = 0;
+    if (!buckets.empty()) {
+        std::optional<size_t> open_index;
+        for (size_t i = 0; i < buckets.size(); ++i) {
+            /// There will always be exactly 1 open_window in the result set
+            if (buckets.at(i).is_open()) {
+                vassert(!open_index, "Data serialization was incorrect");
+                open_index = i;
+                break;
+            }
+        }
+        vassert(open_index, "Data serialization was incorrect");
+        /// Optimization to begin picking up if wall time is within
+        /// window interval
+        const auto begin = std::chrono::seconds(buckets.at(*open_index).begin);
+        const auto now = std::chrono::seconds(epoch_time_secs());
+        vassert(begin < now, "Clock or data serialization error");
+        const auto delta = now - begin;
+        if (delta < _usage_window_width_interval) {
+            /// Continue from where the program left off
+            _current_window = *open_index;
+            last_window_delta = delta;
+        } else {
+            /// Close window and open a new one
+            buckets.at(*open_index).end = epoch_time_secs();
+            _current_window = (*open_index + 1) % buckets.size();
+            buckets.at(_current_window).reset(ss::lowres_system_clock::now());
+        }
+    }
+    _buckets = std::move(buckets);
+    return last_window_delta;
+}
+
+/// Operated on only by shard-0 this prevents execution of background work
+/// spawned by the configuration property update callbacks after stop has been
+/// called on the usage_manager
+static thread_local ss::gate bg_future_gate;
+
+usage_manager::usage_manager(ss::sharded<storage::api>& storage)
+  : _usage_enabled(config::shard_local_cfg().enable_usage.bind())
+  , _usage_num_windows(config::shard_local_cfg().usage_num_windows.bind())
+  , _usage_window_width_interval(
+      config::shard_local_cfg().usage_window_width_interval_sec.bind())
+  , _usage_disk_persistance_interval(
+      config::shard_local_cfg().usage_disk_persistance_interval_sec.bind())
+  , _storage(storage) {
+    _usage_enabled.watch([this] { (void)reset(); });
+    _usage_num_windows.watch([this] { (void)reset(); });
+    _usage_window_width_interval.watch([this] { (void)reset(); });
+    _usage_disk_persistance_interval.watch([this] { (void)reset(); });
+}
+
+ss::future<> usage_manager::reset() {
+    if (
+      !bg_future_gate.is_closed()
+      && (ss::this_shard_id() == usage_manager_main_shard)) {
```

Comment:
this is confusing: above it's stated about this gate, "Operated on only by shard-0", but here its operated on by the current thread before checking what the shard is. also, can `bg_future_gate` live inside `usage_manager`?

Reply:
Yeah the check should be reversed nice catch. I'm also not super happy with the thread local gate, but I didn't want to put the gate inside usage manager when it would effectively be useless on all shards other then shard-0. Thoughts?

Reply:
>would effectively be useless on all shards other then shard-0. Thoughts?

i would put it into the usage manager. you can even add an accesor to it and then use the `oncore` utility to document and enforce the core rules. thread_local usage should be reserved as a solution for other types of technical problems.

Reply:
removing the bg_gate, resolving